lib/utils/qr_parser.dart

```diff
@@ -0,0 +1,336 @@
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an 'AS IS' BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+import 'package:flutter/foundation.dart';
+import 'package:privacyidea_authenticator/utils/crypto_utils.dart';
+import 'package:privacyidea_authenticator/utils/identifiers.dart';
+import 'package:privacyidea_authenticator/utils/logger.dart';
+import 'package:privacyidea_authenticator/utils/utils.dart';
+
+class QrParser {
+  const QrParser();
+
+  /// The method returns a map that contains all the uri parameters.
+  Map<String, dynamic> parseQRCodeToMap(String uriAsString) {
+    Uri uri = Uri.parse(uriAsString);
+    Logger.info('Barcode is valid Uri:', name: 'parsing_utils.dart#parseQRCodeToMap');
+
+    // TODO Parse crash report recipients
+
+    if (uri.scheme != 'otpauth') {
+      throw ArgumentError.value(
+        'parsing_utils.dart#parseQRCodeToMap',
+        'The uri is not a valid otpauth uri but a(n) [${uri.scheme}] uri instead.',
+      );
+    }
+
+    String type = uri.host;
+    if (equalsIgnoreCase(type, enumAsString(TokenTypes.HOTP)) ||
+        equalsIgnoreCase(type, enumAsString(TokenTypes.TOTP)) ||
+        equalsIgnoreCase(type, enumAsString(TokenTypes.DAYPASSWORD))) {
+      return _parseOtpAuth(uri);
+    } else if (equalsIgnoreCase(type, enumAsString(TokenTypes.PIPUSH))) {
+      return _parsePiAuth(uri);
+    }
+
+    throw ArgumentError.value(
+      'parsing_utils.dart#parseQRCodeToMap',
+      'The token type [$type] is not supported.',
+    );
+  }
+
+  Map<String, dynamic> _parsePiAuth(Uri uri) {
+    // otpauth://pipush/LABELTEXT?
+    // url=https://privacyidea.org/enroll/this/token
+    // &ttl=120
+    // &serial=PIPU0006EF87
+    // &projectid=test-d1231
+    // &appid=1:0123456789012:android:0123456789abcdef
+    // &apikey=AIzaSyBeFSjwJ8aEcHQaj4-isT-sLAX6lmSrvbb
+    // &projectnumber=850240559999
+    // &enrollment_credential=9311ee50678983c0f29d3d843f86e39405e2b427
+    // &apikeyios=AIzaSyBeFSjwJ8aEcHQaj4-isT-sLAX6lmSrvbb
+    // &appidios=1:0123456789012:ios:0123456789abcdef
+
+    Map<String, dynamic> uriMap = {};
+
+    uriMap[URI_TYPE] = uri.host;
+
+    // If we do not support the version of this piauth url, we can stop here.
+    String? pushVersionAsString = uri.queryParameters['v'];
+
+    if (pushVersionAsString == null) {
+      throw ArgumentError.value(uri, 'uri', 'Parameter [v] is not an optional parameter and is missing.');
+    }
+
+    try {
+      int pushVersion = int.parse(pushVersionAsString);
+
+      Logger.info('Parsing push token with version: $pushVersion', name: 'parsing_utils.dart#parsePiAuth');
+
+      if (pushVersion > 1) {
```

Comment:
The maximum supported push version should probably be a constant somewhere. no magic numbers